### PR TITLE
feat: Cookie consent exposure handling

### PR DESCRIPTION
**Pull request by tyiuhc**
### Summary

This update will support consent-aware firing of `$impression` events.

Behavior in different states:
1. `GRANTED`: will fire events as normal - immediately when variant actions are applied
2. `REJECTED`: script will not be run
3. `PENDING`: events will be stored in memory, and only fired when consent status is changed to `GRANTED`

`$impression` events are update to contain a timestamp to indicate when the event occurred, which will override the timestamp generated by the `analytics-browser` upon processing. (Relevant [PR1](https://github.com/amplitude/experiment-js-client/pull/223), [PR2](https://github.com/amplitude/Amplitude-TypeScript/pull/1317)). This is necessary for timestamp accuracy for delayed `$impression` firing during the `PENDING` state.

### Checklist

* [X] Does your PR title have the correct [title format](https://github.com/amplitude/experiment-js-client/blob/main/CONTRIBUTING.md#pr-commit-title-conventions)?
* Does your PR have a breaking change?:  No

### CODE REVIEW

File: packages/experiment-tag/src/exposure/consent-aware-exposure-handler.ts [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+
+  /**
+   * Create a consent-aware wrapper for the track method
+   */
+  private createConsentAwareTrack(
+    originalTrack: (event: ExperimentEvent) => boolean,
+  ) {
+    return (event: ExperimentEvent): boolean => {
+      if (event?.eventProperties) {
+        event.eventProperties.time = new Date().getTime();
```
Comment (stephen-choi-amplitude) [inline packages/experiment-tag/src/exposure/consent-aware-exposure-handler.ts]: Assuming we don't account for timezone, `Date.now()` is preferred.

File: packages/experiment-tag/src/experiment.ts [hunk 2]
```diff
@@ -134,6 +136,10 @@ export class DefaultWebExperimentClient implements WebExperimentClient {
 
     this.storage = new ConsentAwareStorage(this.consentOptions.status);
 
+    this.consentAwareExposureHandler = new ConsentAwareExposureHandler(
+      this.consentOptions.status,
+    );
```
Comment (stephen-choi-amplitude) [inline packages/experiment-tag/src/experiment.ts]: Should this be in line 133 block since `this.consentOptions.status` can be undefined? Similarly, should `consentAwareExposureHandler` be initialized if the user has not opted into use consent aware storage?
Or should we consider `this.consentOptions.status` to be GRANTED if not provided?

File: packages/experiment-tag/test/consent-aware-exposure-handler.test.ts
Comment (stephen-choi-amplitude) [inline packages/experiment-tag/test/consent-aware-exposure-handler.test.ts]: can we add a test case for when consent status is null/undefined?

File: packages/experiment-tag/src/experiment.ts (hunk 2 above)
Comment (stephen-choi-amplitude) [reply packages/experiment-tag/src/experiment.ts]: Oh, I thought this was `this.config.consentOptions` not `this.consentOptions`. I see that the private field is set with the default value.

**Review by stephen-choi-amplitude — APPROVED** [review]
